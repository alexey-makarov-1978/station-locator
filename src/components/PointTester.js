import React from "react";
import PropTypes from "prop-types";

import SuitableStation from "./StationLocator";

function PointTester({ point }) {
  // component checks suitable location for the passed point
  // if power = 0 station object doesnt contain data (null)
  const { station, power } = SuitableStation(point);

  return (
    <div>
      {power > 0
        ? `Best link station for point ${point.X}, ${point.Y} is ${
            station.Location.X
          }, ${station.Location.Y} with power ${power.toFixed(2)}`
        : `No link station within reach for point ${point.X}, ${point.Y}`}
    </div>
  );
}

PointTester.propTypes = {
  point: PropTypes.object.isRequired,
};

export default PointTester;
