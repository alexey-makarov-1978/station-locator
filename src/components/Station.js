// encapsulates the station with methods to calculate the distance to the point
// as well as the power of the station to that point
export default function Station(location, reach) {
  this.Location = location;
  this.Reach = reach;

  this.getDistance = function (point) {
    return Math.sqrt(
      Math.pow(Math.abs(this.Location.X - point.X), 2) +
        Math.pow(Math.abs(this.Location.Y - point.Y), 2)
    );
  };

  this.getPower = function (point) {
    const distance = this.getDistance(point);
    return distance > this.Reach ? 0 : Math.pow(this.Reach - distance, 2);
  };
}
