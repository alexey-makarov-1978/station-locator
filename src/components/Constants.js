import Point from "./Point";
import Station from "./Station";

export const Points = [
  new Point(0, 0),
  new Point(100, 100),
  new Point(15, 10),
  new Point(18, 18),
];

export const Stations = [
  new Station(new Point(0, 0), 10),
  new Station(new Point(20, 20), 5),
  new Station(new Point(10, 0), 12),
];
