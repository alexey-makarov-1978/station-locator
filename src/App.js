import React from "react";
import "./App.css";
import PointTester from "../src/components/PointTester";
import { Points } from "../src/components/Constants";

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <h1>Link Station Locator</h1>
        {Points.map((p, index) => (
          <PointTester key={index} point={p} />
        ))}
      </header>
    </div>
  );
}

export default App;
