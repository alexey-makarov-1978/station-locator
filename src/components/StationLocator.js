import { Stations } from "./Constants";

// locates the most suitable station
export default function GetSuitableStation(point) {
  // calculates the station with max power
  const station = Stations.reduce((res, station) => {
    return res.getPower(point) >= station.getPower(point) ? res : station;
  });

  const power = station.getPower(point);

  // returns both power and station.
  // if power equals 0 that means the station in fact is not valid
  // so we return null
  return { station: power === 0 ? null : station, power };
}
