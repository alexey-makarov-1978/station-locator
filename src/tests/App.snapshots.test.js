import React from "react";
import App from "../App";
import renderer from "react-test-renderer";
import { Points } from "../components/Constants";

it("App matching snapshot for root component", () => {
  const tree = renderer.create(<App points={Points} />);
  expect(tree).toMatchSnapshot();
});
