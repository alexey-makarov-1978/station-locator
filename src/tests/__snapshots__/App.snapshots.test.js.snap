// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matching snapshot for root component 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1>
      Link Station Locator
    </h1>
    <div>
      Best link station for point 0, 0 is 0, 0 with power 100.00
    </div>
    <div>
      No link station within reach for point 100, 100
    </div>
    <div>
      Best link station for point 15, 10 is 10, 0 with power 0.67
    </div>
    <div>
      Best link station for point 18, 18 is 20, 20 with power 4.72
    </div>
  </header>
</div>
`;
